desktop_version/src/KeyPoll.cpp

```diff
@@ -138,6 +140,74 @@ static int changemousestate(
     return timeout;
 }
 
+/* Also used in Input.cpp. */
+void recomputetextboxes(void);
+
+bool cycle_language(bool should_recompute_textboxes)
+{
+    extern KeyPoll key;
+
+    if (game.gamestate == TITLEMODE
+    && game.currentmenuname == Menu::translator_options_cutscenetest)
+    {
+        /* Unfortunately, despite how it may appear to be working, the options
+         * are actually language-specific, and the order could be totally
+         * different between languages too. So we can't cycle in this menu. */
+        music.playef(Sound_CRY);
+        return should_recompute_textboxes;
+    }
+    if (game.translator_cutscene_test)
+    {
+        /* Refuse cycling here for similar reasons, even if it seems like it's
+         * working. The text boxes are based off of the language XML and
+         * could be completely different between languages. */
+        music.playef(Sound_CRY);
+        return should_recompute_textboxes;
+    }
+
+    int i = loc::languagelist_curlang;
+    if (key.keymap[SDLK_LSHIFT])
+    {
+        /* Backwards */
+        i--;
+    }
+    else
+    {
+        /* Forwards */
+        i++;
+    }
+    if (!loc::languagelist.empty())
+    {
+        i = POS_MOD(i, loc::languagelist.size());
```

Comment:
Shifting backwards from Arabic goes to Portuguese (PT) instead of Traditional Chinese, probably because of passing -1 to `POS_MOD`.